File: cpp/933_NumberOfRecentCalls.cpp
class RecentCounter {
public:
    deque<int> request_stack;
    int high_range;
    RecentCounter() {
        request_stack = {};
        high_range = 0;
    }
    
    int ping(int t) {
        high_range = t; // strictly increasing values of t
        while (!request_stack.empty() && high_range - 3000 > request_stack.front())
            request_stack.pop_front();
        request_stack.push_back(t);
        return request_stack.size();
    }
};

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter* obj = new RecentCounter();
 * int param_1 = obj->ping(t);
 */

File: swift/933_NumberOfRecentCalls.swift
class RecentCounter {
    var front: Int = 0
    var calls: [Int] = [Int]()
    var size: Int = 0

    func ping(_ t: Int) -> Int {
        while front < size && calls[front] < t - 3000 {
            front += 1
        }
        size += 1
        calls.append(t)
        return size - front
    }
}

/**
 * Your RecentCounter object will be instantiated and called as such:
 * let obj = RecentCounter()
 * let ret_1: Int = obj.ping(t)
 */
